### PR TITLE
Map STL in HashMap

### PR DESCRIPTION
**Pull request by Izhar03**
## Related Issue
- Addition of MAPS in Containers - STL (#156  )

## Proposed Changes
- I have updated the explanation of Map.cpp as you mentioned that do the changes in H-HashMaps only not in Container STL 
- I have added more study links in my Map.cpp and made a function-based approach to Maps

## Additional Info
- I have only updated the Map.cpp only, I have not worked on the unordered maps or any other maps and ignore the solution screenshot directory, I was working on this file in my other repo not related to gssoc I ran the output there to avoid making a gitignore file, just ignore the path the code is same and the solution is working and fine. 

## Checklist

- [x] ✅ My code follows the code style of this project.
- [x] 📝 My change requires a change to the documentation.
- [x] 🎀 I have updated the documentation accordingly.
- [x] 👀 I have read the [**CONTRIBUTING**](https://github.com/Lakhankumawat/LearnCPP/blob/main/Contributing.md) document.
- [ ] ✨ I have added tests to cover my changes.
- [ ] 🚩 All new and existing tests passed.
- [x] 🌟 ed the repo

## Solution 
- Issue #156 
-I have improved the Map.cpp with my new explanation using a function-based approach and more content to read 

-I made a function to call the printing stuff in Map.cpp , Inserted more documents/resources to read etc  

-This change of mine will improve the cleanliness of the code by making functions and adding resources for a better understanding 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Output Screenshots
| Screenshot #1  | 
| ----------- | ----------- |
| Output 1 as per set input  | 
| 
Map ss solution working
|

### CODE REVIEW

**Comment by Lakhankumawat** [comment]
Hey @Izhar03  you need to follow particular guidelines for your README , please follow
these guidelines [click here](https://github.com/Lakhankumawat/LearnCPP#to-add-programs-using-stl-follow-same-approach-as--algorithms) 
also you can see how your README should look like [see here](https://github.com/Lakhankumawat/LearnCPP/blob/main/S-SortingAlgorithms/README.md) 
for more please go through these [here](https://youtu.be/7UgC3HDETRA)

File: H-HashMap/Map.cpp [hunk 1]
```diff
@@ -1,91 +1,88 @@
...
 
 
-#include<iostream>
-#include<map>
-#include<string>
-using namespace std;
-
-/* -------------Maps maintain a self balancing BST------------- */
-// Maps sorts itself on the basis of keys
+#include<bits/stdc++.h> // A High utility header file generally used in Competitive Programming contains 100's header file
```
Comment (n4i9kita) [inline H-HashMap/Map.cpp:4]: Yep, it is a beautiful header file. 
But, as this cpp file is more about helping someone new learn about STL `Map`, it will be better to mention all the header files that are coming in use in this program.

File: H-HashMap/Map.cpp [hunk 2]
```diff
@@ -1,91 +1,88 @@
...
 
+using namespace std;
 
-int main(){
-map<string,int> m;
+// MAPS is a Data Structre which stores KEY-VALUE Pairs 
+// Ordered Maps use Red Black Trees : an advanced data structre : https://www.geeksforgeeks.org/red-black-tree-set-1-introduction-2/
```
Comment (n4i9kita) [inline H-HashMap/Map.cpp:9]: I'd recommend that you create a readme file, and provide useful links there, and not as comments in cpp file.

File: H-HashMap/Map.cpp [hunk 3]
```diff
@@ -1,91 +1,88 @@
...
 
+map<int,string> m; // Declaration Syntax of Map STL 
 
-//Lets try to update the value of keys
-m[fruit]+=22;
+m[1]="abc"; //O(logn) ALSO DEPENDS ON KEY ALSO ( Refer to RED BLACK TREES ) : TIME COMPLEXITY
+m[5]="cdc";
+m[3]="acd";
 
+// Red Black Trees : It compares other strings length and place it acc to sorted method , READ more on above link
```
Comment (n4i9kita) [inline H-HashMap/Map.cpp:41]: This explanation should go into the readme.

File: H-HashMap/Map.cpp [hunk 4]
```diff
@@ -1,91 +1,88 @@
...
+// string comparison time is added therefore
+// Final TIME COMPLEXITY {s.size()*O(logn)} 
 
-//Iterator can also be defined as map<string,int>:: iterator;
-//it here is iterator
-auto it = m.find(fruit);
+auto it1=m.find(3); // returns iterator of 3 if it of 3 doesnt exist then retunr m.end()
```
Comment (n4i9kita) [inline H-HashMap/Map.cpp:45]: Improve the comment.

File: H-HashMap/Map.cpp [hunk 5]
```diff
@@ -1,91 +1,88 @@
...
+  cout<<(*it1).first<<" "<<(*it1).second<<endl;
+}
 
-//If fruit is not present it will point to the end of map
 
-if(it!=m.end()){
-    cout<<"Price of "<<fruit<<" : "<<m[fruit];}
-    else
-        cout<<"Fruit is not present";
+if(it1!=m.end()) // Checking if 'it' exist or not to avoid Segmentation Fault: https://www.geeksforgeeks.org/core-dump-segmentation-fault-c-cpp/
```
Comment (n4i9kita) [inline H-HashMap/Map.cpp:54]: Is this related to the above code?

File: H-HashMap/Map.cpp [hunk 6]
```diff
@@ -1,91 +1,88 @@
...
-    m["Strawberry"]=110;
-    m["Mango"]=90;
+}
 
-    for(auto it=m.begin();it!=m.end();it++)
-        cout<<it->first<<" and "<<it->second<<endl;  //We can also access values by *it
 
 
-    //For each loop in Maps
+// strings are also stored in lexographical order ie dictionary order
```
Comment (n4i9kita) [inline H-HashMap/Map.cpp:79]: Can you please elaborate your comment or explain it using code?
Better be, you can add this point in the readme.

**Comment by Izhar03** [comment]
Thanks, @n4i9kita for such an in-depth constructive review of my code. I will definitely keep this in mind and will work on every single issue. Currently, my exams are going on so please wait till tomorrow, I will resolve all the issues by tomorrow or by the day after tomorrow max.

Actually, I didn't know that with my code I need to update the README.md also that's why I didn't touch it (It's my first time in open source pardon me for that ). I will work on this also as soon as possible.

File: H-HashMap/Map.cpp (hunk 1 above)
Comment (Izhar03) [reply H-HashMap/Map.cpp:9]: Changed and added have a look

File: H-HashMap/Map.cpp (hunk 2 above)
Comment (Izhar03) [reply H-HashMap/Map.cpp]: README.md updated, Have a look please, tried working as per guidelines.

File: H-HashMap/Map.cpp (hunk 5 above)
Comment (Izhar03) [reply H-HashMap/Map.cpp]: Not really I was explaining something else, I removed it this time it's not necessary to explain

**Comment by n4i9kita** [comment]
@Izhar03, it's totally fine. Take your time. 
I really appreciate your desire to learn.


PS : Beware of plagiarism.

**Comment by n4i9kita** [comment]
LGTM

**Comment by Izhar03** [comment]
> LGTM

So is it fine? Or should I work more on it? Is my work done?

**Comment by Lakhankumawat** [comment]
LGTM 💯

**Comment by Lakhankumawat** [comment]
@all-contributors please add @Izhar03  code ,

**Comment by Izhar03** [comment]
Thank you all, Feels excited, This was my first pull request!
Thank you @n4i9kita didi for helping me out throughout, learned a lot from you in the process :)
